internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -137,3 +146,150 @@ func (ar *AlarmsRepository) UpsertAlarmDefinitions(ctx context.Context, records
+			record.ProbableCauseID,
+		)))
+	}
+	query.Apply(values...)
+
+	// Set upsert constraints
+	query.Apply(buildUpsertConstraints())
+
+	return query.Build() //nolint:wrapcheck
+}
+
+// buildUpsertConstraints creates the conflict resolution modifiers
+func buildUpsertConstraints() bob.Mod[*dialect.InsertQuery] {
+	dbTags := utils.GetAllDBTagsFromStruct(models.AlarmDefinition{})
+	return im.OnConflictOnConstraint("unique_fingerprint_alarm_raised_time").DoUpdate(
+		im.SetExcluded(dbTags["AlarmChangedTime"]),
+		im.SetExcluded(dbTags["AlarmClearedTime"]),
+		im.SetExcluded(dbTags["AlarmStatus"]),
+		im.SetExcluded(dbTags["ObjectID"]),
+		im.SetExcluded(dbTags["AlarmDefinitionID"]),
+		im.SetExcluded(dbTags["ProbableCauseID"]),
+	)
+}
+
+// GetAlarmDefinitions needed to build out aer
+func (ar *AlarmsRepository) GetAlarmDefinitions(ctx context.Context, am *api.AlertmanagerNotification, clusterIDToObjectTypeID map[uuid.UUID]uuid.UUID) ([]models.AlarmDefinition, error) {
+	m := models.AlarmDefinition{}
+	dbTags := utils.GetAllDBTagsFromStruct(m)
+	query := psql.Select(
+		sm.Columns(
+			utils.GetColumnsAsAny(utils.GetColumns(m, []string{
+				"AlarmName", "AlarmDefinitionID", "ProbableCauseID",
+				"ObjectTypeID", "Severity",
+			}))...),
+		sm.From(m.TableName()),
+		sm.Where(
+			psql.Group(
+				psql.Quote(dbTags["AlarmName"]),
+				psql.Quote(dbTags["ObjectTypeID"]),
+				psql.Quote(dbTags["Severity"]),
+			).
+				In(getGetAlertNameObjectTypeIDAndSeverity(am, clusterIDToObjectTypeID)...), // Dynamically pass the pairs
+		),
+	)
+
+	sql, params, err := query.Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build alarm definitions query when processing AM notification: %w", err)
+	}
+
+	return utils.ExecuteCollectRows[models.AlarmDefinition](ctx, ar.Db, sql, params)
+}
+
+func getGetAlertNameObjectTypeIDAndSeverity(am *api.AlertmanagerNotification, clusterIDToObjectTypeID map[uuid.UUID]uuid.UUID) []bob.Expression {
+	var b []bob.Expression
+	for _, alert := range am.Alerts {
+		name := alertmanager.GetAlertName(*alert.Labels)
+		_, amSeverity := alertmanager.GetPerceivedSeverity(*alert.Labels)
+		objectTypeId, ok := clusterIDToObjectTypeID[alertmanager.GetClusterID(*alert.Labels)]
+		if !ok {
+			continue
+		}
+
+		b = append(b, psql.ArgGroup(name, objectTypeId, amSeverity))
+	}
+
+	return b
+}
+
+// ResolveNotificationIfNotInCurrent find and only keep the alerts that are available in the current payload
+func (ar *AlarmsRepository) ResolveNotificationIfNotInCurrent(ctx context.Context, am *api.AlertmanagerNotification) error {
+	m := models.AlarmEventRecord{}
+	dbTags := utils.GetAllDBTagsFromStruct(m)
+	query := psql.Update(
+		um.Table(m.TableName()),
+		um.SetCol(dbTags["AlarmStatus"]).ToArg(api.Resolved),
+		um.SetCol(dbTags["AlarmClearedTime"]).ToArg(time.Now()),
+		um.SetCol(dbTags["PerceivedSeverity"]).ToArg(api.CLEARED),
+		um.Where(
+			psql.Group(psql.Quote(dbTags["Fingerprint"]), psql.Quote(dbTags["AlarmRaisedTime"])).
```

Comment:
It seems to me that this is missing a conditional in the Where to only update rows that are not Resolved/Cleared. Otherwise, it will update the AlarmClearedTime with the current time on each notification

Reply:
updated to check AlarmClearedTime is null or not. So only the null fields will have the timestamp.